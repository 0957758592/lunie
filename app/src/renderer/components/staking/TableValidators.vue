<template>
  <div>
    <table class="data-table">
      <thead>
        <panel-sort :sort="sort" :properties="properties" />
      </thead>
      <tbody>
        <li-validator
          v-for="validator in sortedEnrichedValidators"
          :key="validator.operator_address"
          :validator="validator"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import num from "scripts/num"
import { orderBy, isEmpty } from "lodash"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `commission`,
      order: `asc`
    },
    rollingWindow: 10000 // param of slashing period
  }),
  computed: {
    ...mapGetters([
      `delegation`,
      `committedDelegations`,
      `session`,
      `distribution`,
      `liquidAtoms`,
      `bondDenom`,
      `keybase`,
      `pool`
    ]),
    enrichedValidators() {
      return this.validators.map(v =>
        Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          percent_of_vote: v.voting_power / this.pool.pool.bonded_tokens,
          my_delegations: this.committedDelegations[v.id] > 0
            ? this.committedDelegations[v.id]
            : 0,
          commission: v.commission.rate,
          keybase: this.keybase[v.description.identity],
          rewards: this.session.signedIn
            && this.distribution.loaded && !isEmpty(this.distribution.rewards)
            ? this.distribution.rewards[v.operator_address]
            : 0,
          uptime: v.signing_info
            ? (this.rollingWindow - v.signing_info.missed_blocks_counter)
            / this.rollingWindow
            : 0
        })
      )
    },
    sortedEnrichedValidators() {
      return orderBy(
        this.enrichedValidators.slice(0),
        [this.sort.property],
        [this.sort.order]
      )
    },
    properties() {
      return [
        {
          title: `Moniker`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`
        },
        {
          title: `My Delegations`,
          value: `my_delegations`,
          tooltip: `Number of ${
            this.bondDenom
          } you have delegated to this validator`
        },
        {
          title: `Rewards`,
          value: `rewards`,
          tooltip: `Rewards you have earned from this validator`
        },
        {
          title: `Voting Power`,
          value: `percent_of_vote`,
          tooltip: `Percentage of voting shares`
        },
        {
          title: `Commission`,
          value: `commission`,
          tooltip: `The validator's commission`
        },
        {
          title: `Uptime`,
          value: `uptime`,
          tooltip: `Ratio of blocks signed within the last 10k blocks`
        },
      ]
    }
  },
  watch: {
  //   address: function() {
  //     this.session.address && this.$store.dispatch(`updateDelegates`)
  //   },
    validators: function(validators) {
      if (!validators || validators.length === 0 || !this.session.signedIn) {
        return
      }
      const yourValidators = validators.filter(
        ({ operator_address }) => operator_address in this.committedDelegations
      )

      this.$store.dispatch(`getRewardsFromAllValidators`, yourValidators)
    }
  },
  mounted() {
    this.$store.dispatch(`getPool`)
  }
}
</script>
