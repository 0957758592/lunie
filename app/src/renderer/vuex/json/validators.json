[
  {
    "id": "john-connor",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "john-connor.tmp",
    "ipAddress": "35.102.58.60",
    "delegators": 7,
    "atoms": 763344,
    "startDate": "2017-05-25",
    "commission": 0.0105,
    "slashes": []
  },
  {
    "id": "agent-janssen",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "",
    "ipAddress": "72.4.172.62",
    "delegators": 58,
    "atoms": 621242,
    "startDate": "2017-04-11",
    "commission": 0.0095,
    "slashes": []
  },
  {
    "id": "janelle-voight",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "voigh-industries.tmp",
    "ipAddress": "23.144.70.72",
    "delegators": 33,
    "atoms": 532029,
    "startDate": "2016-12-28",
    "commission": 0.011,
    "slashes": []
  },
  {
    "id": "dr-silberman",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "",
    "ipAddress": "209.152.16.89",
    "delegators": 17,
    "atoms": 510694,
    "startDate": "2017-05-09",
    "commission": 0.01,
    "slashes": []
  },
  {
    "id": "serena-korgan",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "serenssimo.tmp",
    "ipAddress": "247.67.58.219",
    "delegators": 51,
    "atoms": 406938,
    "startDate": "2017-05-30",
    "commission": 0.02,
    "slashes": []
  },
  {
    "id": "kate-brewster",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "brewster-validator.tmp",
    "ipAddress": "32.226.184.34",
    "delegators": 44,
    "atoms": 375432,
    "startDate": "2017-03-17",
    "commission": 0.09975,
    "slashes": []
  },
  {
    "id": "miles-dyson",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "dyson-vaccums.tmp",
    "ipAddress": "252.108.145.162",
    "delegators": 38,
    "atoms": 346069,
    "startDate": "2017-02-10",
    "commission": 0.005948,
    "slashes": []
  },
  {
    "id": "marcus-wright",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "wright-brothers.tmp",
    "ipAddress": "251.249.98.149",
    "delegators": 46,
    "atoms": 330528,
    "startDate": "2017-01-03",
    "commission": 0.0875,
    "slashes": []
  },
  {
    "id": "kyle-reese",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "reeses-pieces.tmp",
    "ipAddress": "215.216.123.168",
    "delegators": 50,
    "atoms": 307932,
    "startDate": "2017-05-19",
    "commission": 0.0975,
    "slashes": []
  },
  {
    "id": "sarah-connor",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Text string goes here",
    "url": "sarah-connor.tmp",
    "ipAddress": "67.164.35.159",
    "delegators": 29,
    "atoms": 306238,
    "startDate": "2017-04-13",
    "commission": 0.01125,
    "slashes": []
  },
  {
    "id": "machine-01",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "7.247.140.151",
    "delegators": 13,
    "atoms": 299604,
    "startDate": "2017-04-24",
    "commission": 0.0128,
    "slashes": []
  },
  {
    "id": "machine-02",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "validator-machines.tmp",
    "ipAddress": "9.160.179.15",
    "delegators": 8,
    "atoms": 286434,
    "startDate": "2017-03-18",
    "commission": 0.01024,
    "slashes": []
  },
  {
    "id": "machine-03",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "56.183.206.90",
    "delegators": 65,
    "atoms": 270349,
    "startDate": "2017-02-23",
    "commission": 0.0808,
    "slashes": []
  },
  {
    "id": "machine-04",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "10.239.7.140",
    "delegators": 34,
    "atoms": 262301,
    "startDate": "2017-05-26",
    "commission": 0.0707,
    "slashes": []
  },
  {
    "id": "machine-05",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "five-nines.tmp",
    "ipAddress": "139.19.4.30",
    "delegators": 68,
    "atoms": 256105,
    "startDate": "2017-04-20",
    "commission": 0.0888,
    "slashes": []
  },
  {
    "id": "t-x",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "236.142.154.129",
    "delegators": 45,
    "atoms": 254183,
    "startDate": "2017-04-03",
    "commission": 0.0901,
    "slashes": []
  },
  {
    "id": "guardian-01",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "67.27.236.231",
    "delegators": 73,
    "atoms": 235029,
    "startDate": "2017-03-29",
    "commission": 0.009,
    "slashes": []
  },
  {
    "id": "t-850",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "altered-validator.tmp",
    "ipAddress": "247.248.219.156",
    "delegators": 24,
    "atoms": 183200,
    "startDate": "2017-02-16",
    "commission": 0.096,
    "slashes": []
  },
  {
    "id": "t-800",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "",
    "ipAddress": "87.49.43.91",
    "delegators": 11,
    "atoms": 162395,
    "startDate": "2017-05-08",
    "commission": 0.01108,
    "slashes": []
  },
  {
    "id": "t-1000",
    "pub_key": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDZ67wzRdjbTb9HxduU9YQd9",
    "description": "Specific note about this machine",
    "url": "latest-validator.tmp",
    "ipAddress": "170.19.167.106",
    "delegators": 30,
    "atoms": 84235,
    "startDate": "2017-05-22",
    "commission": 0.094,
    "slashes": []
  }
]
