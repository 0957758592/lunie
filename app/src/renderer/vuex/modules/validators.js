export default ({ commit, node }) => {
  const state = {
    blockchainName: '',
    validators: {},
    validatorHash: null
  }

  const mutations = {
    setValidatorBlockchainName (state, name) {
      state.blockchainName = name
    },
    setValidators (state, validators) {
      state.validators = validators
    },
    setValidatorHash (state, validatorHash) {
      state.validatorHash = validatorHash
    }
  }

  const actions = {
    maybeUpdateValidators (state, header) {
      let validatorHash = header.validators_hash
      if (validatorHash === state.validatorHash) return
      commit('setValidatorHash', validatorHash)
      getValidators()
    }
  }

  function getValidators () {
    node.rpc.validators((err, { validators }) => {
      if (err) return console.error('error fetching validator set')
      commit('setValidators', validators)
    })
  }
  getValidators()

  return { state, mutations, actions }
}
