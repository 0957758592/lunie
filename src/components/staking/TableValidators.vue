<template>
  <div>
    <table class="data-table">
      <thead>
        <PanelSort
          :sort="sort"
          :properties="properties"
          :show-on-mobile="showOnMobile"
        />
      </thead>
      <tbody v-infinite-scroll="loadMore" infinite-scroll-distance="400">
        <LiValidator
          v-for="validator in showingValidators"
          :key="validator.operator_address"
          :validator="validator"
          :show-on-mobile="showOnMobile"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters, mapState } from "vuex"
import num from "scripts/num"
import orderBy from "lodash.orderby"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
import BN from "bignumber.js"
import { expectedReturns } from "scripts/returns"
export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({
    num: num,
    query: ``,
    sort: {
      property: `commission`,
      order: `asc`
    },
    showing: 15,
    rollingWindow: 10000 // param of slashing period
  }),
  computed: {
    ...mapState([`delegates`, `distribution`, `pool`, `session`]),
    ...mapState({
      annualProvision: state => state.minting.annualProvision
    }),
    ...mapGetters([`committedDelegations`, `bondDenom`, `lastHeader`]),
    enrichedValidators(
      {
        validators,
        delegates: { signingInfos },
        pool,
        annualProvision,
        committedDelegations,
        session,
        distribution,
        rollingWindow
      } = this
    ) {
      return validators.map(v => {
        const signingInfo = signingInfos[v.operator_address]
        return Object.assign({}, v, {
          small_moniker: v.description.moniker.toLowerCase(),
          my_delegations:
            session.signedIn && committedDelegations[v.operator_address] > 0
              ? committedDelegations[v.operator_address]
              : 0,
          commission: v.commission.rate,
          voting_power: BN(v.tokens)
            .div(pool.pool.bonded_tokens)
            .toFixed(10),
          rewards:
            session.signedIn && distribution.rewards[v.operator_address]
              ? distribution.rewards[v.operator_address][this.bondDenom]
              : 0,
          uptime: signingInfo
            ? (rollingWindow - signingInfo.missed_blocks_counter) /
              rollingWindow
            : 0,
          expectedReturns: annualProvision
            ? expectedReturns(
                v,
                parseInt(pool.pool.bonded_tokens),
                parseFloat(annualProvision)
              )
            : undefined
        })
      })
    },
    sortedEnrichedValidators() {
      return orderBy(
        this.enrichedValidators.slice(0),
        [this.sort.property],
        [this.sort.order]
      )
    },
    showingValidators() {
      return this.sortedEnrichedValidators.slice(0, this.showing)
    },
    properties() {
      return [
        {
          title: `Name`,
          value: `small_moniker`,
          tooltip: `The validator's moniker`
        },
        {
          title: `Rewards`,
          value: `expectedReturns`,
          tooltip: `Approximate annualized reward`
        },
        {
          title: `Voting Power`,
          value: `voting_power`,
          tooltip: `Percentage of voting shares`
        }
      ]
    }
  },
  watch: {
    lastHeader: {
      immediate: true,
      handler() {
        this.$store.dispatch(`getRewardsFromMyValidators`)
      }
    },
    "sort.property": function() {
      this.showing = 15
    },
    "sort.order": function() {
      this.showing = 15
    }
  },
  mounted() {
    this.$store.dispatch(`getPool`)
    this.$store.dispatch(`getRewardsFromMyValidators`)
    this.$store.dispatch(`getMintingParameters`)
  },
  methods: {
    loadMore() {
      this.showing += 10
    }
  }
}
</script>
<style scoped>
@media screen and (max-width: 550px) {
  .data-table td {
    overflow: hidden;
  }

  .data-table__row__info {
    max-width: 22rem;
  }
}
</style>
