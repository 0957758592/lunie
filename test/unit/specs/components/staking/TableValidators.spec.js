import { shallowMount } from "@vue/test-utils"
import TableValidators from "renderer/components/staking/TableValidators"
import validators from "../../store/json/validators.js"

describe(`TableValidators`, () => {
  let wrapper, $store

  const getters = {
    delegation: {
      loaded: true
    },
    committedDelegations: {
      [validators[2].operator_address]: 10
    },
    session: {
      address: `address1234`
    },
    liquidAtoms: 42,
    connected: true,
    bondDenom: `stake`,
    keybase: {
      [validators[0].description.identity]: {
        // TODO
      }
    },
    pool: {
      pool: {
        bonded_tokens: 500001,
      }
    }
  }

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: JSON.parse(JSON.stringify(getters)) // clone so we don't overwrite by accident
    }

    wrapper = shallowMount(TableValidators, {
      mocks: {
        $store
      },
      propsData: { validators }
    })
  })

  it(`has the expected html structure`, async () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should sort the delegates by selected property`, () => {
    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `desc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address))

    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `asc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address).reverse())
  })

  it(`queries delegations on signin`, () => {
    const session = { address: `cosmos1address` }
    const $store = { dispatch: jest.fn() }
    TableValidators.watch.address.call({ $store, session })
    expect($store.dispatch).toHaveBeenCalledWith(`updateDelegates`)
  })

  it(`doesn't query delegations if not signed in`, () => {
    const session = { address: undefined }
    const $store = { dispatch: jest.fn() }
    TableValidators.watch.address.call({ $store, session })
    expect($store.dispatch).not.toHaveBeenCalledWith(`updateDelegates`)
  })

  xit(`updates if there are existing validators`, () => {
    const session = { signedIn: true }
    $store = {
      dispatch: jest.fn(),
      getters: {
        committedDelegations: {
          [validators[2].operator_address]: 10
        }
      }
    }
    TableValidators.watch.validators.call({ $store, session }, validators)
    expect($store.dispatch).toHaveBeenCalledWith(
      `getRewardsFromAllValidators`,
      validators[2]
    )
  })

  it(`doesn't update rewards if user hasn't signed in`, () => {
    const validators = []
    const session = { signedIn: false }
    TableValidators.watch.validators.call({ $store, session }, validators)
    expect($store.dispatch).not.toHaveBeenCalledWith(
      `getRewardsFromAllValidators`
    )
  })

  it(`doesn't update rewards if validator set is empty`, () => {
    const validators = []
    const session = { signedIn: true }
    TableValidators.watch.validators.call({ $store, session }, validators)
    expect($store.dispatch).not.toHaveBeenCalledWith(
      `getRewardsFromAllValidators`
    )
  })

  it(`doesn't update rewards if validator set is undefined`, () => {
    const validators = undefined
    const session = { signedIn: true }
    TableValidators.watch.validators.call({ $store, session }, validators)
    expect($store.dispatch).not.toHaveBeenCalledWith(
      `getRewardsFromAllValidators`
    )
  })
})
