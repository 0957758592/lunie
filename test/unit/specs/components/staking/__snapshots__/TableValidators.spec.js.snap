// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableValidators shows a validator table 1`] = `
<div>
  <table
    class="data-table"
  >
    <thead>
      <panelsort-stub
        properties="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        sort="[object Object]"
      />
    </thead>
     
    <tbody>
      <livalidator-stub
        validator="[object Object]"
      />
      <livalidator-stub
        validator="[object Object]"
      />
      <livalidator-stub
        validator="[object Object]"
      />
    </tbody>
  </table>
</div>
`;
